warehouse/cli/tuf.py

```diff
@@ -0,0 +1,178 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import click
+
+from securesystemslib.exceptions import StorageError  # type: ignore
+from securesystemslib.interface import (  # type: ignore
+    generate_and_write_ed25519_keypair,
+)
+
+from warehouse.cli import warehouse
+from warehouse.packaging.utils import render_simple_detail
+from warehouse.tuf.tasks import (
+    TargetFile,
+    add_hashed_targets as _add_hashed_targets,
+    bump_bin_n_roles as _bump_bin_n_roles,
+    bump_snapshot as _bump_snapshot,
+    init_dev_repository as _init_dev_repository,
+    init_targets_delegation as _init_targets_delegation,
+)
+
+
+@warehouse.group()  # pragma: no-branch
+def tuf():
+    """
+    Manage Warehouse's TUF state.
+    """
+
+
+@tuf.group()
+def dev():
+    """
+    TUF Development purposes commands
+    """
+
+
+@dev.command()
+@click.pass_obj
+@click.option(
+    "--name", "name_", required=True, help="The name of the TUF role for this keypair"
+)
+@click.option(
+    "--path",
+    "path_",
+    required=True,
+    help="The basename of the Ed25519 keypair to generate",
+)
+def keypair(config, name_, path_):
+    """
+    Generate a new TUF keypair.
+    """
+    password = config.registry.settings[f"tuf.{name_}.secret"]
+    generate_and_write_ed25519_keypair(password, filepath=path_)
+
+
+@dev.command()
+@click.pass_obj
+def init_repo(config):
+    """
+    Initialize a new TUF repository if it does not exist.
+    """
+    try:
+        request = config.task(_init_dev_repository).get_request()
+        config.task(_init_dev_repository).run(request)
+    except FileExistsError as err:
+        raise click.ClickException(str(err))
+
+    click.echo("Repository Initialization finished.")
+
+
+@dev.command()
+@click.pass_obj
+def bump_snapshot(config):
+    """
+    Bump Snapshot metadata.
+    """
+    request = config.task(_bump_snapshot).get_request()
+    config.task(_bump_snapshot).run(request)
+    click.echo("Snapshot bump finished.")
+
+
+@dev.command()
+@click.pass_obj
+def bump_bin_n_roles(config):
+    """
+    Bump delegated targets roles (BIN-N).
+    """
+    request = config.task(_bump_bin_n_roles).get_request()
+    config.task(_bump_bin_n_roles).run(request)
+    click.echo("BIN-N roles (hash bins) bump finished.")
+
+
+@dev.command()
+@click.pass_obj
+def init_delegations(config):
+    """
+    Create delegated targets roles (BINS and BIN-N).
+
+    Given an initialized (but empty) TUF repository, create the delegated
+    targets role (bins) and its hashed bin delegations (each bin-n).
+    """
+    request = config.task(_init_targets_delegation).get_request()
+    try:
+        config.task(_init_targets_delegation).run(request)
+    except (FileExistsError, StorageError) as err:
+        raise click.ClickException(str(err))
+
+    click.echo("BINS and BIN-N roles targets delegation finished.")
+
+
+@dev.command()
+@click.pass_obj
+def add_all_packages(config):
```

Comment:
Running this locally, I got:

```
Traceback (most recent call last):
  File "/opt/warehouse/src/warehouse/tuf/services.py", line 139, in get
    file_object = open(filename, "rb")
FileNotFoundError: [Errno 2] No such file or directory: '/var/opt/warehouse/tuf_metadata/1.bins.json'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/runpy.py", line 196, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/local/lib/python3.10/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/opt/warehouse/src/warehouse/__main__.py", line 18, in <module>
    sys.exit(warehouse())
  File "/opt/warehouse/lib/python3.10/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/opt/warehouse/lib/python3.10/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/opt/warehouse/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/warehouse/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/warehouse/lib/python3.10/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/opt/warehouse/lib/python3.10/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/opt/warehouse/lib/python3.10/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/opt/warehouse/lib/python3.10/site-packages/click/decorators.py", line 38, in new_func
    return f(get_current_context().obj, *args, **kwargs)
  File "/opt/warehouse/src/warehouse/cli/tuf.py", line 148, in add_all_packages
    config.task(_add_hashed_targets).run(request, targets)
  File "/opt/warehouse/src/warehouse/tasks.py", line 71, in run
    result = original_run(*args, **kwargs)
  File "/opt/warehouse/src/warehouse/tuf/tasks.py", line 58, in add_hashed_targets
    repository_service.add_hashed_targets(targets)
  File "/opt/warehouse/src/warehouse/tuf/services.py", line 428, in add_hashed_targets
    bin_n = self._load(RoleType.BINS.value)
  File "/opt/warehouse/src/warehouse/tuf/services.py", line 219, in _load
    return Metadata.from_file(role_name, None, self._storage_backend)
  File "/opt/warehouse/lib/python3.10/site-packages/tuf/api/metadata.py", line 233, in from_file
    with storage_backend.get(filename) as file_obj:
  File "/usr/local/lib/python3.10/contextlib.py", line 135, in __enter__
    return next(self.gen)
  File "/opt/warehouse/src/warehouse/tuf/services.py", line 142, in get
    raise StorageError(f"Can't open {filename}")
securesystemslib.exceptions.StorageError: Can't open /var/opt/warehouse/tuf_metadata/1.bins.json
```

Reply:
Did you run `make init_tuf` before trying to add targets?